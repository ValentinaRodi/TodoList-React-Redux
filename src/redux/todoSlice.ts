import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Todo } from "../models/Todo";
import { v4 as uuidv4 } from "uuid";

const initialState = [] as Todo[];

const todoSlice = createSlice({
  name: "todos",
  initialState,
  reducers: {
    addTodo: {
      reducer: (state: any[], action: PayloadAction<Todo>) => {
        state.push(action.payload);
      },
      prepare: (description: string) => ({
        payload: {
          id: uuidv4(),
          description,
          styleBtn: "item__btn",
          styleText: "item__text",
          counter: 1,
          itemShow: 2,
        } as Todo,
      }),
    },
    removeTodo(state: any[], action: PayloadAction<string>) {
      const index = state.findIndex((todo) => todo.id === action.payload);
      state.splice(index, 1);
    },
    setTodoStatus(
      state: any[],
      action: PayloadAction<{ id: string }>
    ) {
      const index = state.findIndex((todo) => todo.id === action.payload.id);
      state[index].styleBtn = "item__btn_check";
      state[index].styleText = "item__text_check";
      state[index].counter = 0;
    },
    showItemActive(state: any[], action: PayloadAction<{ todoItemShow: number }>) {
      state.map((item)=>item.itemShow = action.payload.todoItemShow)
    },
    removeTodoAll() {
      return initialState
    },
    
  },
});

export const { addTodo, removeTodo, setTodoStatus, showItemActive, removeTodoAll } = todoSlice.actions;
export default todoSlice.reducer;


