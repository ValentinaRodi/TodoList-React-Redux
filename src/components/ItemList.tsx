import { AppDispatch, RootState } from "../redux/store";
import { useDispatch, useSelector } from "react-redux";
import { removeTodo, setTodoStatus } from "../redux/todoSlice";
import { v4 as uuidv4 } from "uuid";
import "./itemList.css";

export default function ItemList() {
  const todoList = useSelector((state: RootState) => state);
  const dispatch = useDispatch<AppDispatch>();

  return (
    <div>
      {todoList.map((todo) => {
        if (todo.itemShow === 2 || todo.itemShow === todo.counter) {
          return (
            <div className="item" key={uuidv4()}>
              <button
                className={todo.styleBtn}
                onClick={() => {
                  dispatch(setTodoStatus({ id: todo.id }));
                }}
              >
                <span>
                  <svg
                    fill="#FFFFFF"
                    xmlns="http://www.w3.org/2000/svg"
                    viewBox="0 0 64 64"
                    width="15px"
                    height="15px"
                  >
                    <path d="M27 55L6 33 9 29 26 41 55 12 59 16z" />
                  </svg>
                </span>
              </button>
              <p className={todo.styleText}>{todo.description}</p>
              <button
                className="item__btn_del"
                onClick={() => {
                  dispatch(removeTodo(todo.id));
                }}
              >
                â•³
              </button>
            </div>
          );
        } else {
          return null;
        }
      })}
    </div>
  );
}
