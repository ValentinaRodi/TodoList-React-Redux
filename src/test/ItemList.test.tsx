import  ItemList  from '../components/ItemList';
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../redux/store";

describe("<ItemList />", () => {
  it("should render successfully and match snapshot", () => {
    const { container } = render(
        <Provider store={store}> 
            <ItemList />
        </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
