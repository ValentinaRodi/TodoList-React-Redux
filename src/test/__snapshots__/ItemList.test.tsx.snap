// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemList /> should render successfully and match snapshot 1`] = `
<div>
  <div />
</div>
`;
